oe_contact_forms.module

```diff
@@ -6,3 +6,327 @@
+ *
+ * @param array $form
+ *   The form.
+ * @param Drupal\Core\Form\FormStateInterface $form_state
+ *   The form state.
+ */
+function _oe_contact_forms_add_one_callback(array &$form, FormStateInterface $form_state): void {
+  $num_topics = $form_state->get('num_topics');
+  $num_topics++;
+  $form_state->set('num_topics', $num_topics);
+  $form_state->setRebuild();
+}
+
+/**
+ * Add one more element to the form based on ajax.
+ *
+ * @param array $form
+ *   The form.
+ * @param Drupal\Core\Form\FormStateInterface $form_state
+ *   The form state.
+ */
+function _oe_contact_forms_remove_callback(array &$form, FormStateInterface $form_state): void {
+  $num_topics = $form_state->get('num_topics');
+
+  if ($num_topics > 1) {
+    $num_topics--;
+  }
+
+  $form_state->set('num_topics', $num_topics);
+  $form_state->setRebuild();
+}
+
+/**
+ * Entity builder for the contact form edit form with third party options.
+ *
+ * @param string $entity_type
+ *   The contact form entity.
+ * @param \Drupal\contact\ContactFormInterface $contact_form
+ *   The contact form entity.
+ * @param array $form
+ *   The form.
+ * @param Drupal\Core\Form\FormStateInterface $form_state
+ *   The form state.
+ */
+function _oe_contact_forms_contact_form_builder($entity_type, ContactFormInterface $contact_form, array &$form, FormStateInterface $form_state): void {
+  $is_corporate_form = (boolean) $form_state->getValue('is_corporate_form');
+  $contact_form->setThirdPartySetting('oe_contact_forms', 'is_corporate_form', $is_corporate_form);
+
+  if ($is_corporate_form === TRUE) {
+    $contact_form->setThirdPartySetting('oe_contact_forms', 'topic_label', $form_state->getValue('topic_label'));
+    $contact_form->setThirdPartySetting('oe_contact_forms', 'email_subject', $form_state->getValue('email_subject'));
+    $contact_form->setThirdPartySetting('oe_contact_forms', 'header', $form_state->getValue('header'));
+    $contact_form->setThirdPartySetting('oe_contact_forms', 'privacy_policy', $form_state->getValue('privacy_policy'));
+    $contact_form->setThirdPartySetting('oe_contact_forms', 'includes_fields_in_auto_reply', $form_state->getValue('includes_fields_in_auto_reply'));
+    $contact_form->setThirdPartySetting('oe_contact_forms', 'allow_canonical_url', $form_state->getValue('allow_canonical_url'));
+    $contact_form->setThirdPartySetting('oe_contact_forms', 'expose_as_block', $form_state->getValue('expose_as_block'));
+    $contact_form->setThirdPartySetting('oe_contact_forms', 'optional_fields', $form_state->getValue('optional_fields'));
+
+    $topics_fieldset = $form_state->getValue('topics_fieldset');
+    $topic_names = [];
+    $topic_email_addresses = [];
+
+    foreach ($topics_fieldset['group'] as $topic) {
+      $topic_names[] = $topic['topic_name'];
+      $topic_email_addresses[] = $topic['topic_email_address'];
+    }
+
+    $contact_form->setThirdPartySetting('oe_contact_forms', 'topic_name', $topic_names);
```

Comment:
The storage of the topic name/email sets is not great. Why don't we store them as one map of topic and email addresses? Like this we rely on the order in the array.

Don't forget to update also the schema to reflect this.

Reply:
Refactored.